### PR TITLE
feat(config): validate `proxy_app`

### PR DESCRIPTION
**Pull request by melekes**
Closes https://github.com/cometbft/cometbft/issues/2719

---

#### PR checklist

- [x] Tests written/updated
- [ ] ~~Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)~~
- [ ] ~~Updated relevant documentation (`docs/` or `spec/`) and code comments~~
- [x] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -324,6 +325,37 @@ func (cfg BaseConfig) ValidateBasic() error {
...
+		"noop":
+	default: // network address
+		parts := strings.SplitN(cfg.ProxyApp, "://", 2)
+		if len(parts) != 2 { // TCP address
+			_, err := net.ResolveTCPAddr("tcp", cfg.ProxyApp)
+			if err != nil {
+				return fmt.Errorf("failed to resolve TCP proxy_app %s: %w", cfg.ProxyApp, err)
+			}
+		} else {
+			switch parts[0] { // other protocol
```
Comment (andynog) [inline config/config.go:348]: here's the validation is only in the protocol (`parts[0]`) but would it be possible to have a validation on the address (`parts[1]`) ? at least from a well form perspective ?

File: config/config.go [hunk 2]
```diff
@@ -324,6 +325,37 @@ func (cfg BaseConfig) ValidateBasic() error {
 	default:
 		return errors.New("unknown log_format (must be 'plain' or 'json')")
 	}
+
+	switch cfg.ProxyApp {
+	case "kvstore",
```
Comment (andynog) [inline config/config.go:330]: These values are the ones in `DefaultClientCreator` right ? Would be possible to have a more idiomatic `enum` way to check them? Like

type ProxyAppType int

const (
    KVStore ProxyAppType = iota
    KVStoreUnsync
    E2E
    ...
)

File: config/config_test.go [hunk 3]
```diff
@@ -70,6 +70,15 @@ func TestBaseConfigValidateBasic(t *testing.T) {
 	require.Error(t, cfg.ValidateBasic())
 }
 
+func TestBaseConfigValidateBasic2(t *testing.T) {
```
Comment (andynog) [inline config/config_test.go:73]: Could we also test more examples here ? Like the `default` use case where we assume it's `tcp` ?

**Review by andynog — APPROVED** [review]
lgtm, added a few comments but those are just additional suggested enhancements.

**Comment by cason** [comment]
I think we create the connection with the client in `DefaultClientCreator` from `proxy/client.go`.

Wouldn't make sense to add a kind of validation method there instead of checking this in the `config` package?

I always think in the case we change anything in the `proxy`, who will remember and do the super boring step of updating the `config` package as well?

File: config/config.go (hunk 2 above)
Comment (cason) [reply config/config.go:330]: This can be a constant []string in any case, just check if it matches any of them.

File: config/config.go [hunk 4]
```diff
@@ -324,6 +325,37 @@ func (cfg BaseConfig) ValidateBasic() error {
...
+		"kvstore_connsync",
+		"kvstore_unsync",
+		"persistent_kvstore",
+		"persistent_kvstore_connsync",
+		"persistent_kvstore_unsync",
+		"e2e",
+		"e2e_connsync",
+		"e2e_unsync",
+		"noop":
+	default: // network address
```
Comment (cason) [inline config/config.go:340]: I wonder if we don't have anything on this style inside the `p2p` package. We have to validate addresses there too...

Comment (melekes) [reply config/config.go:340]: 90% is done by the `net.Dial` function.

**Comment by melekes** [comment]
> Wouldn't make sense to add a kind of validation method there instead of checking this in the config package?

This would defeat the purpose of this PR - validate proxy_app early. There are little to no checks in proxy package. net.Dial does all of them internally.

> I always think in the case we change anything in the proxy, who will remember and do the super boring step of updating the config package as well?

added comments in the code which might help

**Pull request merged**